### PR TITLE
[HLSL][DXIL] Implement `asdouble` intrinsic

### LINKED ISSUE #99081: Implement the `asdouble` HLSL Function
> NOTE: asdouble is special: there are questions around what the right design is for this. If 64-bit integers are supported then we don't need a specific DXIL operation to do this. We also need to consider what the right thing is for the SPIR-V backend.

- [ ] Implement `asdouble` clang builtin,
- [ ] Link `asdouble` clang builtin with `hlsl_intrinsics.h`
- [ ] Add sema checks for `asdouble` to `CheckHLSLBuiltinFunctionCall` in `SemaChecking.cpp`
- [ ] Add codegen for `asdouble` to `EmitHLSLBuiltinExpr` in `CGBuiltin.cpp`
- [ ] Add codegen tests to `clang/test/CodeGenHLSL/builtins/asdouble.hlsl`
- [ ] Add sema tests to `clang/test/SemaHLSL/BuiltIns/asdouble-errors.hlsl`
- [ ] Create the `int_dx_asdouble` intrinsic in `IntrinsicsDirectX.td`
- [ ] Create the `DXILOpMapping` of `int_dx_asdouble` to  `101` in `DXIL.td`
- [ ] Create the  `asdouble.ll` and `asdouble_errors.ll` tests in `llvm/test/CodeGen/DirectX/`

## DirectX

| DXIL Opcode | DXIL OpName | Shader Model | Shader Stages |
| ----------- | ----------- | ------------ | ------------- |
| 101 | MakeDouble | 6.0 | () |

## SPIR-V

There is no support for `asdouble` when targeting SPIR-V.

### Example 1
```hlsl
//dxc asdouble_test.hlsl -T lib_6_8 -enable-16bit-types -O0

export double4 fn(uint4 p1, uint4 p2) {
    return asdouble(p1, p2);
}
```
## HLSL:

Reinterprets a cast value (two 32-bit values) into a double.

## Syntax

``` syntax
double asdouble(
  in uint lowbits,
  in uint highbits
);
```

## Parameters

<dl> <dt>

*lowbits* \[in\]
</dt> <dd>

Type: **uint**

The low 32-bit pattern of the input value.

</dd> <dt>

*highbits* \[in\]
</dt> <dd>

Type: **uint**

The high 32-bit pattern of the input value.

</dd> </dl>

## Return value

Type: **double**

The input (two 32-bit values) recast as a double.

## Remarks

The following overloaded version is also available:

``` syntax
double2 asdouble(uint2 lowbits, uint2 highbits);
```

If the input value is two 32-bit components, the return type will contain one double. If the input value is four 32-bit components, the return type will contain two doubles. If the input value is a 64-bit type, the returned value will have the same number of components as the input value.

### Minimum Shader Model

This function is supported in the following shader models.

| Shader Model                                                                | Supported |
|-----------------------------------------------------------------------------|-----------|
| [Shader Model 5](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/d3d11-graphics-reference-sm5.md) and higher shader models | yes       |

 

This function is supported in the following types of shaders:

| Vertex | Hull | Domain | Geometry | Pixel | Compute |
|--------|------|--------|----------|-------|---------|
| x      | x    | x      | x        | x     | x       |

 

## See also

<dl> <dt>

[Intrinsic Functions](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/dx-graphics-hlsl-intrinsic-functions.md)
</dt> <dt>

[Shader Model 5](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/d3d11-graphics-reference-sm5.md)
</dt> </dl>

### PR DESCRIPTION
**Pull request by inbelic**
- define intrinsic as builtin in Builtins.td
    - link intrinsic in hlsl_intrinsics.h
    - add semantic analysis to SemaHLSL.cpp
    - lower to `llvm` or a `dx` intrinsic when applicable in CGBuiltin.cpp
    - define DXIL intrinsic in IntrinsicsDirectX.td
    - add DXIL op and mapping in DXIL.td
    - enable scalarization of intrinsic

    - add basic sema checking to asdouble-errors.hlsl
    
 Resolves #99081

### CODE REVIEW

**Comment by inbelic** [comment]
Please ignore the first commit when reviewing. It is separately reviewed here: #114849

File: clang/lib/Sema/SemaHLSL.cpp [hunk 1]
```diff
@@ -1870,6 +1870,23 @@ bool SemaHLSL::CheckBuiltinFunctionCall(unsigned BuiltinID, CallExpr *TheCall) {
...
+
+    // Set the return type to be a scalar or vector of same length of double
+    ASTContext &Ctx = SemaRef.getASTContext();
+    auto *VTy = TheCall->getArg(0)->getType()->getAs<VectorType>();
+
+    QualType ResultType =
+        VTy ? Ctx.getVectorType(Ctx.DoubleTy, VTy->getNumElements(),
+                                VTy->getVectorKind())
+            : Ctx.DoubleTy;
+    TheCall->setType(ResultType);
```
Comment (joaosaffran) [inline clang/lib/Sema/SemaHLSL.cpp:1887]: Can we use `SetElementTypeAsReturnType` here ?

File: clang/test/SemaHLSL/BuiltIns/asdouble-errors.hlsl [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+double test_too_few_arg_1(uint p0) {
+  return __builtin_hlsl_asdouble(p0);
+  // expected-error@-1 {{too few arguments to function call, expected 2, have 1}}
+}
+
+double test_too_many_arg(uint p0) {
+  return __builtin_hlsl_asdouble(p0, p0, p0);
+  // expected-error@-1 {{too many arguments to function call, expected 2, have 3}}
+}
```
Comment (joaosaffran) [inline clang/test/SemaHLSL/BuiltIns/asdouble-errors.hlsl:16]: I feel like adding some tests to check for type mismatch would be useful

Comment (inbelic) [reply clang/test/SemaHLSL/BuiltIns/asdouble-errors.hlsl:16]: Is there a type that you have in mind? Most types will be implicitly converted here, I added the custom struct type to illustrate that.

Comment (joaosaffran) [reply clang/test/SemaHLSL/BuiltIns/asdouble-errors.hlsl:16]: Any type really, what we really care about here is making sure the error is meaningful to the user. I usually write such tests with `half`, `float` or `bool` if the former are implicitly cast.

**Review by spall — APPROVED** [review]
LGTM

**Comment by farzonl** [comment]
LGTM, but won't sign off until the `TargetTransformInfo` issue  on the precursor pr is resolved.

**Pull request merged**